test/graceful_termination.jl

```diff
@@ -136,3 +136,113 @@
+                exit(2)
+                return nothing
+            end
+            sleep(60)
+        end
+        cmd = `supervise.sh $(Base.julia_cmd()) --color=no -e $code`
+        cmd = addenv(cmd, "PATH" => shim_path, "DEPUTY_IPC_DIR" => deputy_ipc_dir)
+        buffer = IOBuffer()
+        p = run(pipeline(cmd; stdout=buffer, stderr=buffer); wait=false)
+
+        @test timedwait(() -> process_running(p), Second(5)) === :ok
+
+        # Allow some time for Julia to startup and the graceful terminator to be registered.
+        sleep(3)
+
+        kill(p, Base.SIGTERM)
+        @test timedwait(() -> process_exited(p), Second(10)) === :ok
+        # we exit(2) in the julia code above
+        @test p.exitcode == 2
+
+        output = String(take!(buffer))
+        expected = """
+            [ Info: GRACEFUL TERMINATION HANDLER
+            [ Info: SHUTDOWN COMPLETE
+            """
+        @test contains(output, expected)
+        println(output)
+    end
+
+    @testset "graceless termination" begin
+        code = quote
+            sleep(60)
+        end
+        cmd = `supervise.sh $(Base.julia_cmd()) --color=no -e $code`
+        cmd = addenv(cmd, "PATH" => shim_path, "DEPUTY_IPC_DIR" => deputy_ipc_dir)
+        buffer = IOBuffer()
+        p = run(pipeline(cmd; stdout=buffer, stderr=buffer); wait=false)
+
+        @test timedwait(() -> process_running(p), Second(5)) === :ok
+
+        # Allow some time for Julia to startup and the graceful terminator to be registered.
+        sleep(3)
+
+        kill(p, Base.SIGINT)
+        @test timedwait(() -> process_exited(p), Second(10)) === :ok
+        # SIGTERM=15 gets sent to badly behaving children
+        @test p.exitcode == 15 + 128
+
+        output = String(take!(buffer))
+        expected = r"Child [0-9]+ still running at exit, cleaning up"
+        @test contains(output, expected)
+        println(output)
+    end
+
+    @testset "handles SIGINT" begin
+        code = quote
+            using K8sDeputy
+            atexit(() -> @info "SHUTDOWN COMPLETE")
+            graceful_terminator() do
+                @info "GRACEFUL TERMINATION HANDLER"
+                exit(2)
+                return nothing
+            end
+            sleep(60)
+        end
+        cmd = `supervise.sh $(Base.julia_cmd()) --color=no -e $code`
+        cmd = addenv(cmd, "PATH" => shim_path, "DEPUTY_IPC_DIR" => deputy_ipc_dir)
+        buffer = IOBuffer()
+        p = run(pipeline(cmd; stdout=buffer, stderr=buffer); wait=false)
+
+        @test timedwait(() -> process_running(p), Second(5)) === :ok
+
+        # Allow some time for Julia to startup and the graceful terminator to be registered.
+        sleep(3)
+
+        kill(p, Base.SIGINT)
+        @test timedwait(() -> process_exited(p), Second(10)) === :ok
+
+        # what should this be?  no idea.
+        @test p.exitcode != 0
+
+        output = String(take!(buffer))
+        println(output)
+    end
```

Comment:
funny, these run locally (both on Ubuntu/amd64 and macOS/arm64)

Reply:
er, succeed.

Reply:
I am a fool, they do NOT pass on Ubuntu (I forgot to check out this branch 🤦 )

Reply:
ooookay I can reproduce the failures on macOS as well using a non-ancient version of bash.  seems like the handling of exit codes changed between bash 3 and 5